src/transformers/models/canine/modeling_canine.py

```diff
@@ -1169,7 +1169,9 @@ def forward(
         # Contextualize character embeddings using shallow Transformer.
         # We use a 3D attention mask for the local attention.
         # `input_char_encoding`: shape (batch_size, char_seq_len, char_dim)
-        char_attention_mask = self._create_3d_attention_mask_from_input_mask(input_ids, attention_mask)
+        char_attention_mask = self._create_3d_attention_mask_from_input_mask(
+            input_ids if input_ids is not None else inputs_embeds, attention_mask
+        )
```

Comment:
we could use the `input_shape` defined above, but would require to change this function as well. Let's just go with this. 
I'm pretty sure we have tests in our CI for that make sure generation work with use cache and in that case, you need this to work. Good for me as is